sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -2267,6 +2267,156 @@ case class FormatNumber(x: Expression, d: Expression)
   override def prettyName: String = "format_number"
 }
 
+object ToNumber {
+  def transfer(input: UTF8String, pattern: UTF8String): UTF8String = {
+    val inputStr = input.toString
+    val patternStr = pattern.toString.toUpperCase(Locale.ROOT).replaceAll("FM", "")
+    val hasSign = inputStr.startsWith("-") ||
+      (patternStr.contains('S') && inputStr.contains('-')) ||
+      (patternStr.contains("MI") && inputStr.contains('-')) ||
+      (patternStr.contains("PR") && inputStr.startsWith("<"))
+    val (newInputStr, newPatternStr) = if (hasSign) {
+      (inputStr.replaceAll("-|<|>", ""), patternStr.replaceAll("S|(MI)|(PR)|(FM)", ""))
+    } else {
+      (inputStr, patternStr.replaceAll("FM", ""))
+    }
+    val inputChars = newInputStr.toCharArray()
+    val patternChars = newPatternStr.toIterator
```

Comment:
Why do you need to convert it to iterator? String's `foreach` can iterate over chars as well.

Reply:
Yes. I should avoid the convert.